K8s auto-instrumentation details and example (Java) (#54)
elastic/opentelemetry: docs/kubernetes/operator/instrumenting-java.md

# Instrumenting Java applications with EDOT SDKs on Kubernetes

This document focuses on instrumenting Java applications on Kubernetes, using the OpenTelemetry Operator, EDOT Collectors and the [Elastic Distribution of OpenTelemetry Java](https://github.com/elastic/elastic-otel-java) SDK (EDOT Java).

- For general knowledge about the EDOT Java SDK, refer to the [getting started guide](https://github.com/elastic/elastic-otel-java/blob/main/docs/get-started.md).
- For Java auto-instrumentation specifics, refer to [OpenTelemetry Operator Java auto-instrumentation](https://opentelemetry.io/docs/kubernetes/operator/automatic/#java)
- For general information about instrumenting applications on kubernetes, refer to [instrumenting applications](./instrumenting-applications.md)

## Java agent extensions consideration

> The operator supports configuration that installs [Java agent extensions](https://opentelemetry.io/docs/zero-code/java/agent/extensions/) in `Instrumentation` objects. The extension needs to be available in an image. Refer to [using extensions with the OpenTelemetry Java agent](https://www.elastic.co/observability-labs/blog/using-the-otel-operator-for-injecting-elastic-agents#using-an-extension-with-the-opentelemetry-java-agent) for an example of adding an extension to an agent.

Reviewer: ```suggestion
The operator supports a configuration that installs [Java agent extensions](https://opentelemetry.io/docs/zero-code/java/agent/extensions/) in `Instrumentation` objects. The extension needs to be available in an image. Refer to [using extensions with the OpenTelemetry Java agent](https://www.elastic.co/observability-labs/blog/using-the-otel-operator-for-injecting-elastic-agents#using-an-extension-with-the-opentelemetry-java-agent) for an example of adding an extension to an agent.
```

elastic-instrumentation   107s   http://opentelemetry-kube-stack-daemon-collector.opentelemetry-operator-system.svc.cluster.local:4318
```
> [!NOTE]
> If your `Instrumentation` object has a different name or is created in a different namespace, you will have to adapt the annotation value in the next step.

2. Enable auto-instrumentation of your Java application using one of the following methods:

  - Edit your application workload definition and include the annotation under `spec.template.metadata.annotations`:

    ```yaml
    kind: Deployment
    metadata:
      name: java-app
      namespace: java-ns
    spec:
    ...
      template:
        metadata:
    ...
          annotations:
            instrumentation.opentelemetry.io/inject-java: opentelemetry-operator-system/elastic-instrumentation
    ...
    ```

  - Alternatively, add the annotation at namespace level to apply auto-instrumentation in all Pods of the namespace:

    ```bash
    kubectl annotate namespace java-ns instrumentation.opentelemetry.io/inject-java=opentelemetry-operator-system/elastic-instrumentation
    ```

3. Restart application

>   Once the annotation has been set, the Pods need to be recreated for the instrumentation libraries to be injected.

Author: I'm going to change it to something like "`restart the application` to force the recreation of the Pods" or something like that, as there's no such action like `recreate` the Pods.

The idea is that either the user deletes Pods (the workload controller creates new Pods) or restart the application through `kubectl rollout restart`.

    ```bash
    kubectl rollout restart deployment java-app -n java
    ```

4. Verify the [auto-instrumentation resources](./instrumenting-applications.md#how-auto-instrumentation-works) are injected in the Pod:

>   Run a `kubectl describe` of one of your application pods and check:

Reviewer: ```suggestion
  Run a `kubectl describe` of one of your application Pods and check:
```

    ...
    ```

  - The Pod has an `EmptyDir` volume named `opentelemetry-auto-instrumentation-java` mounted in both the main and the init containers in path `/otel-auto-instrumentation-java`.

    ```bash
    Init Containers:
      opentelemetry-auto-instrumentation-java:
    ...
        Mounts:
          /otel-auto-instrumentation-java from opentelemetry-auto-instrumentation-java (rw)
    Containers:
      java-app:
    ...  
        Mounts:
          /otel-auto-instrumentation-java from opentelemetry-auto-instrumentation-java (rw)
    ...
    Volumes:
    ...
      opentelemetry-auto-instrumentation-java:
        Type:        EmptyDir (a temporary directory that shares a pod's lifetime)
    ```

  Ensure the environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` points to a valid endpoint and there's network communication between the Pod and the endpoint.

5. Confirm data is flowing through in **Kibana**:

  - Open Observability -> Applications -> Service Inventory, and determine if:
    - The application appears in the list of services.
    - The application shows transactions and metrics.
  

>   - For application container logs, open **Kibana Discover** and filter for your pods log. In the provided example we could filter them with any of:

Reviewer: ```suggestion
  - For application container logs, open **Kibana Discover** and filter for your Pod's logs. In the provided example, we could filter for them with either of the following:
```
Not sure if this should be Pods' or Pod's. Is it one Pod or multiple?

For general information about instrumenting applications on kubernetes, refer to [instrumenting applications](./instrumenting-applications.md)

## Guided example to instrument a Java app with EDOT Java SDK on Kubernetes

In the following example you will learn how to:

- Deploy an example Java app in a dedicated namespace.
- Enable auto-instrumentation of the application following any of the supported methods, such as:
  - Adding an annotation to the namespace.
  - Adding an annotation to the deployment pods.
- Verify that auto-instrumentation libraries are injected and configured correctly.
- Confirm data is flowing to **Kibana Observability**.

Before continuing, ensure you have performed the [installation of the operator](./README.md), and confirm that the following `Instrumentation` object exists in the system:

```bash
$ kubectl get instrumentation -n opentelemetry-operator-system
NAME                      AGE    ENDPOINT                                                                                                
elastic-instrumentation   107s   http://opentelemetry-kube-stack-daemon-collector.opentelemetry-operator-system.svc.cluster.local:4318
```

> Auto-instrumentation steps:

Reviewer 2: ```suggestion
Example auto-instrumentation steps:
```

          JAVA_TOOL_OPTIONS:                      -javaagent:/otel-auto-instrumentation-java/javaagent.jar
          OTEL_SERVICE_NAME:                     java-app
          OTEL_EXPORTER_OTLP_ENDPOINT:           http://opentelemetry-kube-stack-daemon-collector.opentelemetry-operator-system.svc.cluster.local:4318
    ...
    ```

  - The Pod has an `EmptyDir` volume named `opentelemetry-auto-instrumentation-java` mounted in both the main and the init containers in path `/otel-auto-instrumentation-java`.

    ```bash
    Init Containers:
      opentelemetry-auto-instrumentation-java:
    ...
        Mounts:
          /otel-auto-instrumentation-java from opentelemetry-auto-instrumentation-java (rw)
    Containers:
      java-app:
    ...  
        Mounts:
          /otel-auto-instrumentation-java from opentelemetry-auto-instrumentation-java (rw)
    ...
    Volumes:
    ...
      opentelemetry-auto-instrumentation-java:
        Type:        EmptyDir (a temporary directory that shares a pod's lifetime)
    ```

  Ensure the environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` points to a valid endpoint and there's network communication between the Pod and the endpoint.

> 4. Confirm data is flowing through in **Kibana**:

Reviewer 2: There can be an intermediate step which is to confirm that the endpoint (the collector or apm server) is getting the data, but that's more detailed than might be worth putting in here

Author: @jackshirazi , either here or in the dedicated `troubleshooting` doc, that will definitely be important. Thanks for reminding it.

how would you verify that exactly? Are there specific logs in the DaemonSet collector (the current endpoint) about the reception and export of traces? or would you look into the application logs for that? I haven't dig yet into the logs details, but it's something we will have to address soon.

Reviewer 2: I haven't checked yet. I think you can set the collector to output the requests to log/console, that's probably how I'd check

Author: That approach should better fit in a troubleshooting doc than here (making note of it for the next troubleshooting doc update). In case there's a straight forward way to determine that traces are flowing with the default installation configuration I'll add it at a later stage.
Thanks!

> <!--> Should we add some screenshots here? <-->

Reviewer 2: screenshots get out of date quickly
